app/src/main/kotlin/org/cru/godtools/ui/dashboard/lessons/LessonsViewModel.kt

```diff
@@ -1,25 +1,107 @@
 import org.cru.godtools.analytics.model.OpenAnalyticsActionEvent.Companion.ACTION_OPEN_LESSON
 import org.cru.godtools.analytics.model.OpenAnalyticsActionEvent.Companion.SOURCE_LESSONS
+import org.cru.godtools.base.Settings
+import org.cru.godtools.db.repository.LanguagesRepository
 import org.cru.godtools.db.repository.ToolsRepository
+import org.cru.godtools.db.repository.TranslationsRepository
+import org.cru.godtools.model.Language
+import org.cru.godtools.model.Language.Companion.filterByDisplayAndNativeName
+import org.cru.godtools.ui.dashboard.tools.ToolsScreen.Filters.Filter
 import org.greenrobot.eventbus.EventBus
 
+const val KEY_SAVED_LESSON_LANGUAGE_LOCALE = "savedLessonLanguageLocale"
+
 @HiltViewModel
-class LessonsViewModel @Inject constructor(private val eventBus: EventBus, toolsRepository: ToolsRepository) :
-    ViewModel() {
-    val lessons = toolsRepository.getLessonsFlow()
-        .map { it.filterNot { it.isHidden }.sortedBy { it.defaultOrder }.mapNotNull { it.code } }
+class LessonsViewModel @Inject constructor(
+    @ApplicationContext
+    private val context: Context,
+    private val eventBus: EventBus,
+    toolsRepository: ToolsRepository,
+    languagesRepository: LanguagesRepository,
+    private val translationsRepository: TranslationsRepository,
+    private val savedStateHandle: SavedStateHandle,
+    val settings: Settings,
+) : ViewModel() {
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val selectedLanguage = savedStateHandle.getStateFlow(KEY_SAVED_LESSON_LANGUAGE_LOCALE, settings.appLanguage)
+        .flatMapLatest { locale ->
+            languagesRepository.findLanguageFlow(locale).map { it ?: Language(locale) }
+        }
+
+    fun updateSelectedLanguage(newLanguage: Language) {
+        savedStateHandle.set(KEY_SAVED_LESSON_LANGUAGE_LOCALE, newLanguage.code)
+    }
+
+    var query = MutableStateFlow("")
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val lessons = selectedLanguage
+        .flatMapLatest { toolsRepository.getLessonsFlowByLanguage(it.code) }
+        .map {
+            it.filterNot { it.isHidden }
+                .sortedBy { it.defaultOrder }
+                .mapNotNull { it.code }
+        }
         .stateIn(viewModelScope, SharingStarted.WhileSubscribed(), emptyList())
 
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val lessonTranslations = toolsRepository.getLessonsFlow()
+        .map { it.mapNotNullTo(mutableSetOf()) { it.code } }
+        .distinctUntilChanged()
+        .flatMapLatest { translationsRepository.getTranslationsFlowForTools(it) }
+
+    private val toolCountsFlow = lessonTranslations
+        .map { translations ->
+            translations
+                .groupBy { it.languageCode }
+                .mapValues { it.value.distinctBy { it.toolCode }.count() }
+        }
+
+    private val lessonsWithAllTranslations = lessons.combine(lessonTranslations) { lessons, translations ->
```

Comment:
I think the name of the property could be clearer. right now it sounds like you are associating all translations with a lesson. Maybe `lessonTranslationPairs`.

This data structure could also be simplified to just a list of translation objects because the lessonCode is available on the Translation object under `toolCode`

Reply:
oh, I see what you were doing here, the variable named `translation` confused me since it made it seem that it was dealing with a single `translation`.

Reply:
this property is only used by `lessonsWithTranslation`, so maybe combine it to a single property that is more straightforward